Ready for review: revisions to ch 10

carols10cents opened the pull request:
Whew.

It's probably easier to just read the new text rather than the diff at this point; the diff is pretty noisy. I rearranged a lot of stuff.

In addition to the comments, this should fix:

* #411 
* #410 
* #369 
* #368 
* #348 
* #335 

Also I snuck in references to long cat, horse_ebooks, and moby dick. they'll all probably come out in copy edit, but i am pleased with myself for the moment.

## REVIEWERS AND POTENTIAL REVIEWERS

I need a break from this chapter, so we shipped it to nostarch, but we'll have another round of edits with them in a few weeks, so please open new issues if there's anything confusing or wrong or missing!!! ❤️

carols10cents commented on src/ch10-01-syntax.md:
```diff
@@ -292,90 +317,188 @@ enum Result<T, E> {
 </figure>
 
-Compiling this definition results in an error:
+This code isn't particularly useful aside from illustrating that generic types
```
This example is almost but not quite the definition of `Option::and`. I don't really want to make it into `Option::and`, because then we're implementing a method directly on `Option` and also taking an `Option` as an argument, I feel like that might be too many options :(

ScottAbbey replied:
This section is hard to understand and feels out of place next to the others which are all straightforward with simple justification.

carols10cents replied:
Yep, I'm struggling to come up with useful code that doesn't have too many distractions from demonstrating generics in method signatures that are different than generics in struct fields. And also don't need trait bounds. You don't happen to have any ideas, do you? :)

ScottAbbey replied:
I think the biggest problem is explaining this one takes several examples.  If we do something like this, we can show the full extent with just one example.

How about using Point, since it's already introduced in this section.

```
struct Point<T, U> {
    x: T,
    y: U,
}

impl<T, U> Point<T, U> {
    fn mixup<V, X>(self, other: Point<V, X>) -> Point<T, X> {
        Point {
            x: self.x,
            y: other.y,
        }
    }
}

fn main() {
    let p1 = Point { x: 5, y: 10 };
    let p2 = Point { x: "Hello", y: "World"};
    
    let p3 = p1.mixup(p2);
    
    println!("p3.x = {}, p3.y = {}", p3.x, p3.y);
}
```

carols10cents replied:
Love it. I'll incorporate it tomorrow. Thank you thank you thank you <3

carols10cents commented on src/ch10-02-traits.md:
```diff
@@ -1,306 +1,492 @@
+`Vec` in our `aggregator` crate, since we've defined `Summarizable` there. This
+restriction is part of what's called the *orphan rules*, which you can look up
+if you're interested in type theory. Briefly, [INSERT SENTENCE ABOUT WHY IT'S
+CALLED THE ORPHAN RULE AND WHY WE DON'T WANT ORPHAN IMPLS HERE]
```
@steveklabnik I added this paragraph because of #335. I want to mention this, and that it's called orphan rule(s?) so that people can look up more about it, but i feel like we need a sentence that explains why it's called orphan and why it's necessary. I don't know if that's possible though. wdyt?

carols10cents commented:
[insert a "10 things i hate about you" style poem here about how much i hate this chapter here, including ending in denial of hatred and crying here] 😭😭😭😭😭😭😭

carols10cents commented:
oh I deployed this to my fork's gh pages for more convenient pretty reading: https://carols10cents.github.io/book/ch10-00-generics.html

matthewjasper commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -291,23 +319,162 @@ fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
+straightforward example that should match your intuition from any language:
+`string1` is valid until the end of the outer scope, `string2` is valid until
+the end of the inner scope, and `result` is also valid until the end of the
+outer scope. The borrow checker approves of this code; it will compile and
```
I think this is supposed to mean that `result` references something that is valid until the end of the outer scope. 
But then '`string1` is valid' and '`result` is valid' don't appear to use the same definition of valid: the former refers to how long the variable can be referenced, the latter to how long the thing it references is valid for.

carols10cents replied:
If I change this text to:

> `string1` is valid until the end of the outer scope, `string2` is valid until the end of the inner 
> scope, and `result` references something that is valid until the end of the outer scope

does that resolve the issue?

matthewjasper replied:
Yes, this is better.

matthewjasper commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -291,23 +319,162 @@ fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
-doing. If the function didn't actually return the longest string slice but
-instead always returned the first parameter, we wouldn't need to specify a
-lifetime on `y`. This code compiles:
+doing. For example, if we changed the implementation of the `largest` function
```
s/largest/longest/

matthewjasper commented on src/ch10-03-lifetime-syntax.md:
````diff
@@ -359,17 +534,60 @@ rather than a reference so that the calling function is then responsible for
+<figure>
+<span class="filename">Filename: src/main.rs</span>
+
+```
````
Should be '```rust'

matthewjasper commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -359,17 +534,60 @@ rather than a reference so that the calling function is then responsible for
+
+fn main() {
+    let novel = String::from("Call me Ishmael. Some years ago...");
+    let first_sentence = novel.split('.')
```
Is it worth another example with Iterators before Ch13. I guess it's clear enough.

carols10cents replied:
Yeahhhhh maybe I should make this a slice of elements from a `Vec` instead of a string slice-- I'm doing this to avoid slicing on character indices :-/ I kind of wanted to go with string slice here, since trying to put a string slice in a struct is often how people hit this, buuuut....

carols10cents replied:
I think I've decided to send this example to our editor and see what she says.

matthewjasper commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -386,108 +604,279 @@ fn first_word(s: &str) -> &str {
-    name: String,
-    config: &'a Config,
+impl<'a> Important<'a> {
+    fn first_word_of_part(&self) -> &str {
```
But I would write `fn first_word_of_part(&self) -> &'a str` as it's slightly more general. Admittedly it's hard to find examples where it matters apart from for the sake of even more lifetimes.
Using a struct that contains a reference probably complicates this more than is needed.

carols10cents replied:
You're saying you wouldn't elide the lifetime here, even though you're allowed to...?

carols10cents replied:
I also don't tend to store references in structs, but this is often one of the first things beginners try... I wonder if there's an example from "learning rust with entirely too many linked lists" that I could use in here....

carols10cents replied:
Wait wait wait are you saying that you'd add a lifetime annotation in the following case, even though you don't need to, or are you saying that you'd only add the lifetime annotation in the case that the struct has the lifetime?


```
struct Person {
   name: String,
}

impl Person {
    fn first_initial(&self) -> &str {
        &self.name[0..1]
    }
}
```

matthewjasper replied:
I'm saying that elision would make
`fn first_word_of_part<'b>(&'b self) -> &'b str` 
while
`fn first_word_of_part(&self) -> &'a str`, explicitly:
`fn first_word_of_part<'b>(&'b self) -> &'a str`
is more general, the struct doesn't need to stay borrowed while using the string slice. Again, it's surprisingly hard to find (good, simple) examples where the second allows more code. The best I can come up with is getting a `&'static str` back from an `Important<'static>`.

Regarding LRWTMLL: [Iter](http://cglab.ca/~abeinges/blah/too-many-lists/book/second-iter.html) shows this off both a case where this is used
```
fn next(&mut self) -> Option<&'a T>
```
Except that this isn't what's written, there's an associated type that hides it.

In that `Person` case elision does the only possible thing.

matthewjasper commented on src/ch10-03-lifetime-syntax.md:
````diff
@@ -359,17 +534,60 @@ rather than a reference so that the calling function is then responsible for
+<span class="filename">Filename: src/main.rs</span>
+
+```
+struct Important<'a> {
````
Could use a more suggestive name.

carols10cents replied:
How about `ImportantExcerpt`? Have any other suggestions? Naming in examples is hard :(

steveklabnik commented on src/ch10-01-syntax.md:
```diff
@@ -1,203 +1,179 @@
-more of their fields. Generic structs also get monomorphized into specialized
-types at compile time. Listing 10-2 shows the definition and use of a `Point`
-struct that could hold `x` and `y` coordinate values that are any type:
+The note mentions `std::cmp::PartialOrd`, which is a trait defined by the
```
we might want to foreshadow "we'll talk about traits later" here

carols10cents replied:
specifically in the first sentence, rather than 2 sentences later where i have it?

steveklabnik replied:
my reaction was as i was reading it, so yeah, I think. i don't feel mega mega strongly with it being this close, though

steveklabnik commented on src/ch10-02-traits.md:
```diff
@@ -1,306 +1,492 @@
+`Vec` in our `aggregator` crate, since we've defined `Summarizable` there. This
+restriction is part of what's called the *orphan rules*, which you can look up
+if you're interested in type theory. Briefly, [INSERT SENTENCE ABOUT WHY IT'S
+CALLED THE ORPHAN RULE AND WHY WE DON'T WANT ORPHAN IMPLEMENTATIONS HERE]
```
the TL;DR here is about code breakage; do you want me to provide some text here?

carols10cents replied:
yes please :)

steveklabnik replied:
> Briefly, it's called the "orphan rule" because the parent type is not present. The reason that we have this rule is that without it, two crates could implement the same trait for the same type, and they would conflict. By following the orphan rule, other people's code can't break your code, and vice versa.

numist commented:
As luck would have it, I came across [this comment today on the topic of code duplication](https://www.reddit.com/r/programming/comments/5txp5t/duplication_is_far_cheaper_than_the_wrong/ddq5ykf/?st=iz600o33&sh=11d9d7c4) that does a really good job of describing my discomfort around using deduplication as a code-quality metric.

As much as I appreciate DRYing out code, the target audience for this seems to be people with little to no programming experience, and it would be nice to nudge them more explicitly on the right philosophical path :)

carols10cents commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -1,66 +1,34 @@
+The main aim of lifetimes is to prevent dangling references, which will cause a
+program to reference data other than the data we're intending to reference.
+Consider the program in Listing 10-16, with an outer scope and an inner scope.
+The outer scope declares a variable named `r` with no initial value, and the
```
uuugggghhhhh i'm remembering a time when we did have uninitialized variables in the variable introduction section... and i think i was even the one who advocated getting rid of it, haha.

I can see this being concerning/distracting to someone thinking about rust's safety guarantees and freaking out with "omg uninitialized variables!!!!1!!!11!11!". What I'm hoping they will then immediately do is go try and use an uninitialized variable and see that it's a compile-time error. What I'm hoping they *won't* immediately do is tell their CTO "we can't use rust, it's not safe, it allows uninitialized variables".

I am weighing these possible scenarios against addressing the issue here, revising chapter 3 to address it there, or not addressing it. And i am conflicted and will consider it.

@steveklabnik wdyt?

steveklabnik replied:
I feel like the right place is chapter 3. But we could also do it here, as a sidebar. The downside of doing it in chapter 3 is that we can't show off how smart Rust is, like, that it knows control flow to guarantee that it gets assigned to.

carols10cents commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -142,112 +132,87 @@ a dangling reference, and see what the lifetimes look like:
+and below). If these topics are confusing you in this context, I'd be
+interested to know if rereading Chapter 4 clears up that confusion.
+
+/Carol -->
```
LOLOLOLOOOOMG this hides like half the section from the html version that i deployed and tweeted out... can't believe i haven't gotten any complains yet!! Just deployed a fix.

carols10cents commented on src/ch10-00-generics.md:
````diff
@@ -79,56 +113,87 @@ fn main() {
 ```rust
-fn largest(numbers: Vec<i32>) {
-    let mut largest = numbers[0];
+fn largest(list: &[i32]) -> i32 {
````
do you mean just in the example where we have `largest_char` or do you mean all through, since you commented on this instance?

steveklabnik replied:
I meant for just in the example where we make two of them

carols10cents commented:
@matthewjasper thank you so much for catching the details I missed! I *think* I've addressed all your comments at this point, except the use of iterator... I'm going to leave it in for now and see what our editor thinks.

@numist I totally agree with you that creating abstractions isn't about getting the number of duplicated lines of code down, and that it isn't always bad to have duplicated code. While we don't want this book to recommend bad programming habits, there's a lot of nuance around this (and other) issues of software design that would be distractions from our core goal of getting the reader productive in Rust. So I've tweaked a few sentences to be more about duplicated concepts rather than code, and talked about the benefits of clarity and reasoning that come from abstractions. If there are any parts I missed that still seem like bad recommendations or could use rephrasing, please let me know!

@steveklabnik I think I'm done, want to take one last look?

Anyone in the process of reading and reviewing this, I would still LOVE your thoughts and feedback whenever you have time-- we'll have another chance to make edits to this chapter when it gets to the copy-editing stage in a few weeks. I think this is the 2nd most important chapter after Chapter 4, so I want us to make it as good as possible, and that means I want as much feedback as possible :)

ScottAbbey commented on src/ch10-01-syntax.md:
```diff
@@ -1,203 +1,181 @@
+The note mentions `std::cmp::PartialOrd`, which is a *trait*. We're going to
+talk about traits in the next section, but briefly, what this error is saying
+is that the body of `largest` won't work for all possible types that `T` could
+be: since we want to compare values of type `T` in the body, we can only use
```
I think my brain was expecting a semi-colon.  The colon caused me to re-read the whole sentence.

ScottAbbey commented on src/ch10-01-syntax.md:
```diff
@@ -1,203 +1,181 @@
+`std::cmp::PartialOrd` that types can implement to enable comparisons. We'll
+come back to traits and how to specify that a generic type has a particular
+trait in the next section, but let's set this example aside for a moment and
+explore other places we can use generic type parameters first.
```
I know this follows on from the intro section for the chapter, but it was a bit sad to be left with a non-working example and then have several other examples to work through before getting back to it.

The use of generics with structs and enums seems simpler than with functions, the examples are shorter and easy to explain.  You were able to build them all without needing to mention traits.

Please consider moving the use of generics with functions to the end of this section.

carols10cents replied:
So we used to have it that we went in this order:

- demonstrate extracting a regular old function
- demonstrate extracting an enum with generics
- demonstrate generics in structs
- demonstrate generics in methods
- demonstrate generics in functions, which leads into traits and trait bounds since it's hard to write a generic function that does anything without them

but our editor got really confused and expected the order to be more like what we have now. I'm not sure that she realizes *why* we had it in the order we did, namely the issue that we need trait bounds, and you may have seen my HTML comment in the markdown trying to explain as much. 

I want to see what she says about this ordering-- she has expressed the same concern about leaving the reader with non-working examples as you have here, so we'll be talking about the tradeoffs with her and it might indeed go back to what it was.

carols10cents commented on src/ch10-01-syntax.md:
```diff
@@ -216,74 +194,123 @@ fn main() {
+
+### Using Generic Data Types in Enum Definitions
+
+In a similar way as structs, enums can be defined to hold generic data types in
```
idk, both "similar way as" and similar way to" sound strange to me here. What about "Similarly to structs"?

carols10cents commented on src/ch10-01-syntax.md:
```diff
@@ -292,90 +319,193 @@ enum Result<T, E> {
+
+Another way to use generics in methods is with parameters and return types of
+the methods that don't have to be the same as the generic types in the type
+we're implementing methods on. Listing 10-10 has a struct `A<T>` that holds a
```
Hmm. How about:

> Generic type parameters in a struct definition aren't always the same generic type parameters you want to use in that struct's method signatures.

carols10cents commented on src/ch10-00-generics.md:
```diff
@@ -79,56 +113,91 @@ fn main() {
+the `largest` function, the code actually runs on the specific values that we
+pass in.
+
+The mechanics we went through to get from Listing 10-2 to Listing 10-3 were
```
That could be confusing with "struct methods" though :(

steveklabnik commented:
LGTM 😄

ScottAbbey commented on src/ch10-02-traits.md:
```diff
@@ -1,306 +1,492 @@
+The behavior of a type consists of the methods we can call on that type.
+Different types share the same behavior if we can call the same methods on all
+of those types. Trait definitions are a way to group method signatures together
+in order to define a set of behaviors necessary to accomplish some purpose.
```
Coming back to this one after reading the rest of the section...see if there's anything in here you'd like to use:

> It is convenient to be able to perform the same operation on data of several types.  In Chapter 4, we used the `clone` method to create a deep copy of a `String`.  The `clone` method is also available on other types we are already familiar with, such as `Vec<T>`.  Behavior that is shared across types can be expressed as a trait to tell the compiler that ....[I'm running out of steam]...

[And we can also create our own!]

carols10cents replied:
Yeah I think I was trying to go for defining what we mean by "shared behavior" in this context-- it's not anything, it's "which methods you can call", and traits give a name to a group of related methods. Do you think this paragraph would be more useful as:

> By *shared behavior*, we mean that different types respond to the same messages: they have the same methods with the same signatures. A *trait definition* consists of one or more method signatures grouped together outside of any particular type. A *trait implementation* is when we take the method signatures of the trait and define the bodies of the methods for a particular type. Then we can write code elsewhere that calls the methods of the trait, and that code doesn't need to know anything about the type except that it implements the trait, which the compiler will enforce for us.

carols10cents commented on src/ch10-02-traits.md:
```diff
@@ -1,306 +1,492 @@
+our `aggregator` crate functionality. We could also implement `Summarizable` on
+`Vec` in our `aggregator` crate, since we've defined `Summarizable` there. This
+restriction is part of what's called the *orphan rule*, which you can look up
+if you're interested in type theory. Briefly, it's called the orphan rule
```
Yeah I don't think there is one though :)

carols10cents commented on src/ch10-02-traits.md:
````diff
@@ -1,306 +1,492 @@
-fn show_anything<T: Display>(value: T) {
-    println!("I have something to show you!");
-    println!("It's: {}", value);
+```rust,ignore
````
If we're going to refer back to listings, we give them a number.

ScottAbbey commented on src/ch10-01-syntax.md:
```diff
@@ -375,48 +375,37 @@ we specify that we're implementing methods on the type `Point<T>`.
-The method `some_if_i_have_some` takes a value of a different generic type `U`
-and returns an `Option<U>` based on whether the data in the `value` field of
-the struct is `Some(T)` or `None`:
+10-10 defines a method `mix` on the `Point<T, U>` struct from Listing 10-8. The
```
`mix`->`mixup` (and again 2 lines down)

ScottAbbey commented on src/ch10-02-traits.md:
```diff
@@ -47,12 +47,13 @@ behavior provided by a `summary` method
+describe the behaviors that types that implement this trait will need to have,
+in this case `fn summary(&self) -> String`. After the method signature, instead
+of providing an implementation within curly braces, we put a semicolon. Each
+type that implements this trait can then provide its own custom behavior for
```
maybe `must then provide`, since `can` almost makes it sound optional?

matthewjasper commented on src/ch10-01-syntax.md:
```diff
@@ -292,90 +319,169 @@ enum Result<T, E> {
+type parameters you want to use in that struct's method signatures. Listing
+10-10 defines a method `mix` on the `Point<T, U>` struct from Listing 10-8. The
+method takes another `Point` as a parameter, which might have different types
+than the `self` `Point` that we're calling `mix` on. The method creates a new
```
Both `mix`s should be `mixup`.

carols10cents commented on src/ch10-03-lifetime-syntax.md:
```diff
@@ -124,15 +102,34 @@ annotations:
+<!-- Yes, the inside block for the `'b` lifetime starts with the `let x = 5;`
+line and ends with the first closing curly brace on the 7th line. Do you think
+the text art comments work or should we make an SVG diagram that has nicer
+looking arrows and labels? /Carol -->
```
I'm not sure how these are going to end up being printed.

Timmmm replied:
See also #370

carols10cents commented:
I need a break from this chapter, so I'm sending it in now. Thank you everyone for your amazing feedback!!!! ❤️ If you have more feedback, please file new issues and we'll address them in the next round of edits on this chapter in a few weeks!!
